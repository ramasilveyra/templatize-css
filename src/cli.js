/* eslint-disable no-console */
import yargs from 'yargs';
import chalk from 'chalk';
import ora from 'ora';
import { compileFile } from './index';
import pkg from '../package.json';

export default function templatizeCssCLI(argv) {
  console.log(chalk.bold.white(`${pkg.name} v${pkg.version}`));

  const parsedArgv = yargs(argv)
    .option('o', {
      alias: 'out-file',
      describe: 'Output file',
      type: 'string'
    })
    .usage(`${pkg.description}.\nUsage: $0 <file or dir> [options]`)
    .version()
    .alias('version', 'v')
    .help()
    .alias('help', 'h').argv;
  const file = parsedArgv._[0];
  const spinner = ora({ text: 'Processing...' });
  const progress = report => {
    spinner.stopAndPersist({ text: `${chalk.gray(report)}` });
  };

  if (file && parsedArgv.o) {
    spinner.start();
    return compileFile(file, parsedArgv.o, {
      progress
    })
      .then(() => {
        spinner.succeed(`${chalk.bold.green('success')} CSS templatized`);
      })
      .catch(err => {
        spinner.stopAndPersist();
        spinner.fail(`${chalk.bold.red('error')} ${err.stack}`);
        process.exit(1);
      });
  }

  return null;
}
