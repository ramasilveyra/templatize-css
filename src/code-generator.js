import postcss from 'postcss';
import { camelCase } from 'lodash';

export default function codeGenerator(info) {
  const css = generateCss(info.ast);
  const templateRaw = generateTemplateRaw(css);
  const template = fixLocals(templateRaw, info.tracks);
  return template;
}

function generateCss(ast) {
  let css = '';
  postcss.stringify(ast, i => {
    css += i;
  });
  return css;
}

function generateTemplateRaw(css) {
  const isEmpty = css === '';
  const template = `module.exports = function templatizeCSS(${isEmpty ? '' : 'locals'}) {
  const css = ${isEmpty ? "''" : `\`${css}\``};
  return css;
};
`;
  return template;
}

function fixLocals(templateRaw, tracks) {
  let template = templateRaw;
  tracks.forEach(track => {
    template = template.replace(`var(${track.prop})`, `\${locals.${camelCase(track.prop)}}`);
  });
  return template;
}
