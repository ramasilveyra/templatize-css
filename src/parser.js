import postcss from 'postcss';

export default function parser(css) {
  const ast = postcss.parse(css);
  return ast;
}
