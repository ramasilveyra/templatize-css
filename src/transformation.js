import postcss from 'postcss';
import { camelCase } from 'lodash';

const FOLLOW_MSG = 'templatize-css: track';

export default function transformation(oldAst) {
  const tracks = getTracks(oldAst);
  const newAst = generateAst(oldAst, tracks);
  const info = { tracks, ast: newAst };
  return info;
}

function getTracks(ast) {
  let tracks = [];
  ast.walkComments(comment => {
    if (comment.text.startsWith(FOLLOW_MSG)) {
      tracks = [].concat(tracks).concat([{ startLine: comment.source.start.line }]);
    }
  });

  ast.walkRules(':root', rootRule => {
    rootRule.walkDecls(dec => {
      const followFound = tracks.find(follow => follow.startLine === dec.source.end.line);
      if (followFound) {
        tracks = tracks.map(follow => {
          if (followFound === follow) {
            return { name: dec.prop, defaultValue: dec.value, nameJs: camelCase(dec.prop) };
          }
          return follow;
        });
      }
    });
  });
  return tracks;
}

function generateAst(oldAst, tracks) {
  const newAst = postcss.root();
  oldAst.each(node => {
    appendNode(tracks, newAst, node);
  });
  return newAst;
}

function appendNode(tracks, father, node) {
  switch (node.type) {
    case 'rule': {
      if (node.selector === ':root') {
        return;
      }
      const rule = postcss.rule({ selector: node.selector });
      node.walkDecls(node2 => {
        const isTracked = !!tracks.find(f => node2.value.includes(f.name));
        if (isTracked) {
          rule.append(node2);
        }
      });
      if (rule.nodes && rule.nodes.length > 0) {
        father.append(rule);
      }
      break;
    }
    case 'atrule': {
      const atrule = postcss.atRule({ name: node.name, params: node.params });
      node.each(node2 => {
        appendNode(tracks, atrule, node2);
      });
      if (atrule.nodes && atrule.nodes.length > 0) {
        father.append(atrule);
      }
      break;
    }
    default:
      break;
  }
}
