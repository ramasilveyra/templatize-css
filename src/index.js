import fs from 'fs';
import path from 'path';
import util from 'util';
import makeDir from 'make-dir';
import parser from './parser';
import transformation from './transformation';
import codeGenerator from './code-generator';

const readFile = util.promisify(fs.readFile);
const writeFile = util.promisify(fs.writeFile);

export function compile(input) {
  const ast = parser(input);
  const info = transformation(ast);
  const template = codeGenerator(info);
  return template;
}

export async function compileFile(inputFile, outputFile, { progress = () => {} } = {}) {
  const inputFilePath = path.resolve(process.cwd(), inputFile);
  const outputFilePath = path.resolve(process.cwd(), outputFile);
  const inputCode = await readFile(inputFilePath, { encoding: 'utf8' });
  const template = compile(inputCode);
  const outputFilePathDir = path.dirname(outputFilePath);
  await makeDir(outputFilePathDir);
  await writeFile(outputFilePath, template);
  const relativePath = path.join(process.cwd(), '/');
  progress(`${inputFile.replace(relativePath, '')} -> ${outputFile.replace(relativePath, '')}`);
}
