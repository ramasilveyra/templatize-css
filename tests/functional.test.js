import { compile, compileFile } from '../src/index';
import { getTestCase, getFixturePath, getFixture } from './utils';

describe('while using templatizeCss.compile()', () => {
  it('should convert simple CSS file to js template', async () => {
    const { input, output } = await getTestCase('simple');
    const result = compile(input);
    expect(result).toBe(output);
  });
  it('should convert CSS file with extra decls to js template', async () => {
    const { input, output } = await getTestCase('with-extra-decls');
    const result = compile(input);
    expect(result).toBe(output);
  });
  it('should convert CSS file with multiple rules to js template', async () => {
    const { input, output } = await getTestCase('multiple-rules');
    const result = compile(input);
    expect(result).toBe(output);
  });
  it('should convert CSS file with no var to track to js template', async () => {
    const { input, output } = await getTestCase('no-track');
    const result = compile(input);
    expect(result).toBe(output);
  });
  it('should convert CSS file with media queries to js template', async () => {
    const { input, output } = await getTestCase('media-queries');
    const result = compile(input);
    expect(result).toBe(output);
  });
  it('should convert complex CSS file to js template', async () => {
    const { input, output } = await getTestCase('complex');
    const result = compile(input);
    expect(result).toBe(output);
  });
  it('should convert CSS file with multiple decls to js template', async () => {
    const { input, output } = await getTestCase('multiple-decls');
    const result = compile(input);
    expect(result).toBe(output);
  });
});

describe('while using templatizeCss.compileFile()', () => {
  it('should convert css file to js template', async () => {
    const fixture = 'compile-file';
    const inputPath = getFixturePath(`${fixture}/input.css`);
    const outputPathTemplate = getFixturePath(`${fixture}/template.js`);
    await compileFile(inputPath, outputPathTemplate);
    const outputEJS = await getFixture(`${fixture}/template.js`);
    expect(outputEJS).toMatchSnapshot();
  });
});
