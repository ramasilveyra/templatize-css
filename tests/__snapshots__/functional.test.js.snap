// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`while using templatizeCss.compileFile() should convert css file to js template 1`] = `
"module.exports = function templatizeCSS(locals) {
  const css = \`.btn {
  background-color: \${locals.mainPrimaryColor};
  color: \${locals.mainTextColor}
}
@media (min-width: 992px) {
  .btn::after {
    background-color: \${locals.mainMobilePrimaryColor}
  }
}\`;
  return css;
};
"
`;
